rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+
+By getting both parties to agree to a prescriptive standard, new technological possibilities become viable. The Block Protocol is an example of this, enabling a new class of applications, ones that are able to dynamically load complex front-end components at run-time which are immediately usable and functional, even if the component was previously unknown to the application.
+
+The foundation of this capability is standardizing how requirements are expressed. Most notably, the block and entity schemas expressed through [the graph service](https://blockprotocol.org/docs/spec/graph-service-specification) allow for dynamic resolution of expected input data.
+
+---
+
+**Example**:
+
+A block expresses that it needs a URL hosting an image (presumably so that it can render it):
+
+```json
+{
+  "properties": {
+    "image_url": "string"
+  }
+}
+```
+
+The embedding application is then able to search across its data to find entities that satisfy this requirement, such as:
+
+```json
+{
+  "image_url": "https://dummyimage.com/300.jpg",
+  ... // some other fields
+}
+```
+
+It is then able to pass the value to the block, satisfying its requirements for input data.
+
+---
+
+This works most easily when the data uses the same identifier (key in the object) as the block expects, but problems arise if people have different ways of describing compatible data. For example, the embedding application might have an entity that looks like:
+
+```json
+{
+  "display_picture": "https://dummyimage.com/300.jpg",
+  ... // some other fields
+}
+```
+
+This is an example of a lack of **consensus**, where two parties don't agree on how to describe the data in question. At this point, there isn't a clear route to deciding compatibility. Block schemas could perhaps be modified so that the block could express it needs something that matches a [regex to ensure the URL has an image extension](https://stackoverflow.com/questions/42853961/regex-that-matches-image-links-but-exclude-normal-url), and the embedding application could ignore the key and just match on values, but this:
+
+- is potentially flakey ([here's an example](https://dummyimage.com/300) of a URL that points to an image but _does not_ have an extension)
+- ends up being very domain-specific
+- introduces a large maintenance burden
+- is generally not reusable, possibly leading to similar bespoke logic being defined in multiple places, with possible subtle but incompatible variations
+
+Thus the main reason for following RFC:
+
+> 💬 Removing some flexibility to establish a stronger contract between the block developer, embedding application, and end-user, allows for more powerful run-time inference and general confidence of all parties. Furthermore, creating a well-defined hierarchy of types, that are reusable and discoverable, encourages emergent standardization and, in turn, leads to higher-quality interoperable parts.
```

Comment:
I'm not sure I agree as strongly with this one. People do not have the option of avoiding the Type System under this RFC, as such there is a contract being established that is more constrained (i.e. stronger, perhaps not _better_). Perhaps "stronger contract" has connotations that are too firmly implying direct benefit?

Also regarding the inclusion of "which" I don't _think_ that's grammatically correct, but it's generally because the sentence is very wordy and runs on.
> Removing some flexibility ..., allows for more powerful ...

vs 

> Removing some flexibility ..., which allows for more ....

I think general confidence kind of encompasses more of the subtle benefits of the approach, but perhaps this is too hand-wavy and overplays the benefits. In that case perhaps the general way you're leaning here with the review, of picking more specific key benefits, makes more sense (i.e. in this scenario focusing on improvements in data compatibility). I'm open to suggestion, if you feel remotely strongly I'm happy to commit that :)

Reply:
I don't feel particularly strongly. I think this would be addressed by making the 'a lot of benefits are unlocked by people sharing types, not just introducing this new system' somewhere.

For what it's worth I think there are two sorts of contracts between block and app introduced by the type system:
1. We agree to all define our schemas in this composable/re-usable way (they must do this)
2. Where we refer to the same schema (they might do this), it means the same thing

Reply:
I agree with those points, although I'm not sure how much time we want to spend refining this technical document to capture all of those things (not saying you're suggesting we work them in).

As an action item though we'll add a clear section near the top about how most of the benefits are basically gated by people participating in an ecosystem and sharing things.

Reply:
Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/555dd097384f416a42afacfe001870a98ed096e0, made it short-ish as to not spend too much time refining the contents